### PR TITLE
GH2603: Added .NET Core 3 moniker to Cake.Tool

### LINKED ISSUE #2603: Add .NET Core 3 to Cake.Tool update to .NET Core 3 SDK
* Add netcoreapp3.0 moniker to Cake.Tool
* Execute unit tests under netcoreapp3.0
* Update SDK to .NET Core 3

### PR DESCRIPTION
**Pull request by devlead**
(GH-2603) Added .NET Core 3 moniker to Cake.Tool
* Update SDK to .NET Core 3 (3.0.100)
* Add netcoreapp3.0 moniker to Cake.Tool
* Add netcoreapp3.0 moniker to Unit tests
* Execute unit tests under netcoreapp3.0
* Consolidate Cake.Tool.csproj into Cake.csproj (sorts VS2019 rebuild issues)
* Refactor Runtime Framework detection
* Add runtime compiler constants
* Microsoft.CodeAnalysis.CSharp.Scripting 3.3.1
* Remove mono workaround
* Add .NET Core rollForward policy
* Build script cleanup
* Update NuGet command line to 5.2.0 (5.3.0 not blessed yet)
* Update NuGet client libs to 5.3.0
* Temp fix for rollforward test
* fixes #2603

### CODE REVIEW

File: src/Shared.msbuild [hunk 1]
```diff
@@ -38,6 +39,10 @@
     <DefineConstants>$(DefineConstants);NETCORE</DefineConstants>
     <DebugType>portable</DebugType>
   </PropertyGroup>
+  <PropertyGroup Condition=" '$(TargetFramework)' == 'netcoreapp3.0' ">
```
Comment (mholo65) [inline src/Shared.msbuild]: Can we combine all these above?

**Comment by devlead** [comment]
> I think this PR should include the following, if it’s merged before my PR https://github.com/cake-build/cake/pull/2564/files#r314080758

Will take a look at that, code seems to work but, fails for unit tests as xunit isn't compiled (which is entry assembly) doesn't have to be compiled for same version as Cake.dll, so might need to look at executing runtime another way.

**Comment by devlead** [comment]
@mholo65 did a quick PoC detecting running and not compiled runtime.  And that works...a little too well, because unit tests fail if not all runtimes installed, as 2.0 will run on 2.1,  2.1 will run on 3.0, etc.

Taking entry attribute won't work in unit tests as entry assembly is xunit which is netcoreapp 1.0

So while it probably works correctly at runtime, testing "expects" are calculated at compile time, so this is interesting.

At same time for integration tests it's not wrong to test on all supported runtimes.

**Comment by devlead** [comment]
@cake-build/cake-team please review, there's a unit test that smells, but other than that most kinks have been ironed out to run and compile on .NET Core 3.

File: build.ps1 [hunk 2]
```diff
@@ -50,6 +50,11 @@ if ($PSVersionTable.PSEdition -ne 'Core') {
 # INSTALL .NET CORE CLI
 ###########################################################################
 
+$env:DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+$env:DOTNET_CLI_TELEMETRY_OPTOUT=1
+$env:DOTNET_ROLL_FORWARD_ON_NO_CANDIDATE_FX=1
```
Comment (gep13) [inline build.ps1]:
```suggestion
$env:DOTNET_ROLL_FORWARD_ON_NO_CANDIDATE_FX=2
```

Comment (devlead) [reply build.ps1]: Fixed.

File: src/Shared.msbuild [hunk 3]
```diff
@@ -6,38 +6,32 @@
     <Authors>Patrik Svensson, Mattias Karlsson, Gary Ewan Park, Alistair Chapman, Martin Björkström, Dave Glick, Pascal Berger and contributors</Authors>
     <Company>Patrik Svensson, Mattias Karlsson, Gary Ewan Park, Alistair Chapman, Martin Björkström, Dave Glick, Pascal Berger and contributors</Company>
     <PackageLicenseExpression>MIT</PackageLicenseExpression>
-    <PackageIconUrl>https://raw.githubusercontent.com/cake-build/graphics/master/png/cake-medium.png</PackageIconUrl>
+    <PackageIcon>cake-medium.png</PackageIcon>
```
Comment (gep13) [inline src/Shared.msbuild]: From the experiments that @Jericho has done, I believe that NuGet.org will strip the icon out, as it isn't currently supported there.  Can you confirm @Jericho?

Comment (devlead) [reply src/Shared.msbuild]: .NET Core 3 requires it though or you get an compiler error.
```diff
- error NU5048: The 'PackageIconUrl'/'iconUrl' element is deprecated.
- Learn more at https://aka.ms/deprecateIconUrl
```

Comment (gep13) [reply src/Shared.msbuild]: Hmm, interesting.  So I wonder if there is going to be an update to NuGet.org ahead of the RTM of .Net Core 3.

Comment (gep13) [reply src/Shared.msbuild]: Also, was this previously a warning?  Or has it jumped straight to error?

Comment (devlead) [reply src/Shared.msbuild]: Well currently we treat warnings as errors.

Comment (gep13) [reply src/Shared.msbuild]: Oh, I see.  Ok, so that changes things :smile:

Comment (Jericho) [reply src/Shared.msbuild]: I confirm that NuGet.org strops our embedded icons. I uploaded a package with such icon and you can see here that the icon is now missing: https://www.nuget.org/packages/Cake.Email.Common/

Comment (devlead) [reply src/Shared.msbuild]: Looking at NuGet/NuGetGallery/issues/6907 it looks like it's in milestone [S159 - 2019.09.16](https://github.com/NuGet/NuGetGallery/milestone/130) which is due by October 04, 2019 which feels late as .NET Core 3 is due by next week.

Comment (devlead) [reply src/Shared.msbuild]: @Jericho @gep13 Just pushed https://www.nuget.org/packages/LitJson/0.14.1 with PackageIcon and it worked just fine.

Exchanged emails with nuget team and there is no `<icon>` element in nuspec in Cake.Email.Common 0.4.0.
Issue with Cake.Email.Common 0.4.0 is likely not building using .NET Core 3.0 SDK

Can't find any global.json, .NET Core SDK install in build.ps1/build.sh nor appveyor.yml (which seems to use image: Visual Studio 2017), so can't find anything indicating it using newer required SDK.

Comment (Jericho) [reply src/Shared.msbuild]: @devlead You're probably right: when I built on my machine (where .net 3.0 is available), the generated package included the icon but when built on [AppVeyor](https://ci.appveyor.com/project/cakecontrib/cake-email-common/builds/27254454/artifacts), the icon was stripped and therefore the package published to NuGet was missing an icon. I incorrectly assumed that NuGet had stripped the icon.

Comment (devlead) [reply src/Shared.msbuild]: That's one use case for having a `global.json` file, so same SDK is used locally and CI.

**Comment by devlead** [comment]
@cake-build/cake-team PR rebased, updated to RTM .NET Core SDK 3.0.100 and Microsoft.CodeAnalysis.CSharp.Scripting 3.3.1

File: src/Cake.Core.Tests/Unit/CakeRuntimeTests.cs [hunk 4]
```diff
@@ -26,7 +26,35 @@ public void Should_Return_Correct_Result_For_CoreClr()
...
+                    Assert.Equal(".NETCoreApp,Version=v" +
+
+#if NETCOREAPP2_0
+                    "2.0",
+#elif NETCOREAPP2_1
+                    "2.1",
+#elif NETCOREAPP2_2
+                    "2.2",
+#elif NETCOREAPP3_0
+                                 "3.0",
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/CakeRuntimeTests.cs]: Indentation

File: src/Cake.Core.Tests/Unit/CakeRuntimeTests.cs [hunk 5]
```diff
@@ -26,7 +26,35 @@ public void Should_Return_Correct_Result_For_CoreClr()
...
+#endif
+                        framework.FullName);
+                }
+                catch
+                {
+                    // Temp fix for if only netcore3 installed
+                    if (framework.FullName != ".NETCoreApp,Version=v3.0")
```
[the remark this answers is not in the document]
Comment (devlead) [reply src/Cake.Core.Tests/Unit/CakeRuntimeTests.cs]: Fixed, but would like to refactor that test, perhaps make runtime detection more "mockable" but maybe that could be future improvement.

[the remark this answers is not in the document]
Comment (patriksvensson) [reply src/Cake.Core.Tests/Unit/CakeRuntimeTests.cs]: Yes, that would be good. Agree to push it for now.

File: src/Cake.Core/Polyfill/EnvironmentHelper.cs [hunk 6]
```diff
@@ -119,7 +123,28 @@ public static Runtime GetRuntime()
...
+
+            var assemblyPath = typeof(System.Runtime.GCSettings)?.GetTypeInfo()?.Assembly?.CodeBase;
+            if (string.IsNullOrEmpty(assemblyPath))
+            {
+                return NetStandardFramework;
+            }
+
+            const string microsoftNetCoreApp = "Microsoft.NETCore.App";
+            var runtimeBasePathLength = assemblyPath.IndexOf(microsoftNetCoreApp) + microsoftNetCoreApp.Length + 1;
+            var netCoreAppVersion = string.Concat(assemblyPath.Skip(runtimeBasePathLength).Take(3));
```
Comment (devlead) [inline src/Cake.Core/Polyfill/EnvironmentHelper.cs]: > Could this be explained in a comment? Where is the magic number 22 coming from?

@mholo65 better?

**Comment by patriksvensson** [comment]
Merged! Thank you for your contribution. Much appreciated! 👍

**Pull request merged**